intercom-utils/src/cpp.rs

```diff
@@ -0,0 +1,230 @@
+
+use std::collections::HashMap;
+use parse::*;
+use clap::ArgMatches;
+use intercom_common::utils;
+use intercom_common::guid::*;
+use error::*;
+use std::io;
+
+/// Runs the 'cpp' subcommand.
+pub fn run( cpp_params : &ArgMatches ) -> AppResult {
+
+    // Parse the sources and convert the result into an IDL.
+    let ( renames, result ) = parse_crate(
+            cpp_params.value_of( "path" ).unwrap() )?;
+    result_to_cpp( &result, &renames, &mut io::stdout() );
+
+    Ok(())
+}
+
+/// Converts the parse result into an header  that gets written to stdout.
+fn result_to_cpp(
+    r : &ParseResult,
+    rn : &HashMap<String, String>,
+    out : &mut io::Write,
+) {
+    // Introduce all interfaces so we don't get errors on undeclared items.
+    let itf_forward_declarations = r.interfaces.iter().map(|itf| {
+        format!( "{}struct {};", get_indentation( 1 ), try_rename( rn, &itf.name ) )
+    } ).collect::<Vec<_>>().join( "\n" );
+
+    // Raw interfaces that give direct access to the components implemented in Rust.
+    let itfs = r.interfaces.iter().map(|itf| {
+        generate_raw_interface( rn, itf, 1 )
+    } ).collect::<Vec<_>>().join( "\n" );
+
+    // Generate class descriptors.
+    let class_descriptors = r.class_names.iter().map(|class_name| {
+
+        // Get the class details by matching the name.
+        let class  = r.classes.iter().find(|cls| &cls.name == class_name )
+            .unwrap();
+
+        generate_class_descriptor( rn, &r.libname, class_name, class, 1 )
+    } ).collect::<Vec<_>>().join( "\n" );
+
+    // Generate using statements that flatten the raw interfaces
+    // to mimic the results of a Microsoft MIDL compiler.
+    let flat_interface_declarations = r.interfaces.iter().map(|itf| {
+        flatten_interface( rn, &r.libname, itf, 1 )
+    } ).collect::<Vec<_>>().join( "\n" );
+
+    // Generate using statements that flatten the class descriptors
+    // to mimic the results of a Microsoft MIDL compiler.
+    let flat_class_descriptor_declarations = r.class_names.iter().map(|class_name| {
+
+                // Get the class details by matching the name.
+        let class  = r.classes.iter().find(|cls| &cls.name == class_name )
+            .unwrap();
+
+        flatten_class_descriptor( class_name, class, 1 )
+    } ).collect::<Vec<_>>().join( "\n" );
+
+    // We got the interfaces and classes. We can format and output the raw interfaces.
+    let raw_namespace: String = format!( r###"
+>       #include <array>
+>       #include <intercom.h>
+>       namespace {}
+>       {{
+>       namespace raw
+>       {{
```

Comment:
The main implementation is quite far because the default implementation is does not put the classes into a namespace. Also the "C" style definitions are omitted completely. However, you can define INTERCOM_FLATTEN_DECLARATIONS which should alias the declarations into global namespace.

I would not change this at this point. The target is still a bit fuzzy anyway until we start using this more.